cli/azd/pkg/infra/provisioning/infra_provider.go

```diff
@@ -0,0 +1,83 @@
+package provisioning
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/async"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+)
+
+type InfrastructureProviderKind string
+
+const (
+	Bicep     InfrastructureProviderKind = "Bicep"
+	Arm       InfrastructureProviderKind = "Arm"
+	Terraform InfrastructureProviderKind = "Terraform"
+	Pulumi    InfrastructureProviderKind = "Pulumi"
+)
+
+type InfrastructureOptions struct {
+	Provider InfrastructureProviderKind `yaml:"provider"`
+	Path     string                     `yaml:"path"`
+	Module   string                     `yaml:"module"`
+}
+
+type ProvisionPlanResult struct {
+	Plan ProvisioningPlan
+}
+
+type ProvisionPlanProgress struct {
+	Message   string
+	Timestamp time.Time
+}
+
+type ProvisionApplyResult struct {
+	Operations []tools.AzCliResourceOperation
+	Outputs    map[string]ProvisioningPlanOutputParameter
+}
+
+type ProvisionDestroyResult struct {
+	Resources []tools.AzCliResource
+	Outputs   map[string]ProvisioningPlanOutputParameter
+}
+
+type ProvisionApplyProgress struct {
+	Timestamp  time.Time
+	Operations []tools.AzCliResourceOperation
+}
+
+type ProvisionDestroyProgress struct {
+	Message   string
+	Timestamp time.Time
+}
+
+type InfraProvider interface {
```

Comment:
FWIW - `plan` and `apply` are both terms of art in Terraform - so maybe we should pick different names here for cases where things don't line up.

I think that `Plan` and `UpdatePlan` here are not what we will want on this interface long term.  `Plan` right now is more about getting metadata about the infrastructure (it returns both a set of configuration parameters for the IaC file and a set of outputs that it sets) and `UpdatePlan` is about applying an updated set of configuration.

Perhaps we should drop these two methods and instead add `Configuration` and `Outputs` as methods (which return the two parts of the IaC structure we'll want) and then have `Apply` take a configuration object that has the actual configured parameters?

Reply:
Yes, you are correct... they are very Terraformy at the moment.  It may take a couple provider implementations to figure out the best naming but love the feedback.